<a href="https://colab.research.google.com/github/krupa2002/Data-Science-101-From-Fundamentals-to-Mastery/blob/main/Data_Science_101_From_Fundamentals_to_Mastery.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science 101: From Fundamentals to Mastery Part - 1

For a detailed theoretical explanation and insights, check out the [blog post](https://medium.com/@dharamshikrupa/data-science-101-from-fundamentals-to-mastery-part-1-5123b084ef32) that complements this notebook

## Understanding data

In [ ]:
import pandas as pd #importing libraries
import numpy as np

In [ ]:
data = pd.read_csv("/content/titanic.csv") # storing the data in the dataframe named data.

In [ ]:
data #reading the data

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
0      male  34.5      0      0              330911    7.8292   NaN        Q  
1    female  47.0      1      0              363272    7.0000   NaN        S  
2      male  62.0      0      0              240276    9.6875   NaN        Q  
3      male  27.0      0      0              315154    8.6625   NaN        S  
4    female  22.0      1      1             3101298   12.2875   NaN        S  
..      ...   ...    ...    ...                 ...       ...   ...      ...  
413    male   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  female  39.0      0      0            PC 17758  108.9000  C105        C  
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416    male   NaN      0      0              359309    8.0500   NaN        S  
417    male   NaN      1      1                2668   22.3583   NaN        C  

[418 rows x 11 columns]

In [ ]:
data.head(4) #reading 1st 4 rows

PassengerId  Pclass                              Name     Sex   Age  SibSp  \
0          892       3                  Kelly, Mr. James    male  34.5      0   
1          893       3  Wilkes, Mrs. James (Ellen Needs)  female  47.0      1   
2          894       2         Myles, Mr. Thomas Francis    male  62.0      0   
3          895       3                  Wirz, Mr. Albert    male  27.0      0   

   Parch  Ticket    Fare Cabin Embarked  
0      0  330911  7.8292   NaN        Q  
1      0  363272  7.0000   NaN        S  
2      0  240276  9.6875   NaN        Q  
3      0  315154  8.6625   NaN        S

In [ ]:
data.tail(4) #reading last 4 rows

PassengerId  Pclass                          Name     Sex   Age  SibSp  \
414         1306       1  Oliva y Ocana, Dona. Fermina  female  39.0      0   
415         1307       3  Saether, Mr. Simon Sivertsen    male  38.5      0   
416         1308       3           Ware, Mr. Frederick    male   NaN      0   
417         1309       3      Peter, Master. Michael J    male   NaN      1   

     Parch              Ticket      Fare Cabin Embarked  
414      0            PC 17758  108.9000  C105        C  
415      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416      0              359309    8.0500   NaN        S  
417      1                2668   22.3583   NaN        C

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


In the above details of the data we can see that-
1. Pclass has a value set from 1 to 3 so its a ordinal data.
2. Sex has 2 values male or female so its a binary data.
3. Age has range and has fraction value so its a continuous data.
4. SibSp (number of siblings/spouses aboard) and Parch (number of parents/children aboard) this columns are countable so they come under discrete.

Understanding the types of data, such as ordinal, binary, continuous, and discrete, helps us effectively analyze and interpret the dataset


## Key terms for measure of central tendancy

Mean - The sum of all values divided by the number of values (average)

In [ ]:
# Drop rows with NaN values in relevant columns for calculations
age = data['Age'].dropna()
fare = data['Fare']

In [ ]:
# Calculating mean for Age and fare
mean_age = age.mean()
mean_fare = fare.mean()

In [ ]:
print(f"Mean Age: {mean_age:.2f}")
print(f"Mean Fare: {mean_fare:.2f}")

Mean Age: 30.27
Mean Fare: 35.63


In [ ]:
# Calculating median for Age and fare
median_age = age.median()
median_fare = fare.median()

In [ ]:
print(f"Median Age: {median_age:.2f}")
print(f"Median Fare: {median_fare:.2f}")

Median Age: 27.00
Median Fare: 14.45


In [ ]:
# Calculating mode for Age and fare
mode_age = age.mode()
mode_fare = fare.mode()

In [ ]:
print(f"Mode Age: {', '.join(map(str, mode_age))}")
print(f"Mode Fare: {', '.join(map(str, mode_fare))}")

Mode Age: 21.0, 24.0
Mode Fare: 7.75


In [ ]:
# Percentile Calculation
percentile_age_25 = np.percentile(age, 25)  # 25th percentile
percentile_age_75 = np.percentile(age, 75)  # 75th percentile

In [ ]:
print(f"25th Percentile Age: {percentile_age_25:.2f}")
print(f"75th Percentile Age: {percentile_age_75:.2f}")

25th Percentile Age: 21.00
75th Percentile Age: 39.00


In [ ]:
from scipy.stats import trim_mean
# Trimmed Mean Calculation (10% trimming)
trimmed_mean_age = trim_mean(age, proportiontocut=0.1)
trimmed_mean_fare = trim_mean(fare, proportiontocut=0.1)

In [ ]:
print(f"10% Trimmed Mean Age: {trimmed_mean_age:.2f}")
print(f"10% Trimmed Mean Fare: {trimmed_mean_fare:.2f}")

10% Trimmed Mean Age: 29.64
10% Trimmed Mean Fare: 22.18


From the above data and insights, we can see that -

- Mean Age (30.27) and Mean Fare (35.63): Provides an overall average but can be skewed by outliers.

- Median Age (27.00) and Median Fare (14.45): Offers a central value that is less affected by outliers compared to the mean. This suggests that the data might be skewed since the mean and median values differ.

- Mode Age (21.0, 24.0) and Mode Fare (7.75): Shows the most frequent values. The presence of multiple modes for age indicates some common age groups, while a single mode for fare shows a common fare amount.

- 25th and 75th Percentiles (Age: 21.00, 39.00): Provides insight into the distribution of age values, indicating that 50% of the data falls between these percentiles.

- 10% Trimmed Mean Age (29.64) and Fare (22.18): Adjusts for outliers, providing a more robust measure of central tendency. This shows that trimming has slightly reduced the mean values, indicating the influence of extreme values.


## Key terms for Measure of variability

In [ ]:
variance_age = age.var()


In [ ]:
print(f"Variance Age: {variance_age:.2f}")


Variance Age: 201.11


In [ ]:
std_dev_age = age.std()


In [ ]:
print(f"Standard Deviation Age: {std_dev_age:.2f}")


Standard Deviation Age: 14.18


In [ ]:
range_age = age.max() - age.min()


In [ ]:
print(f"Range Age: {range_age:.2f}")


Range Age: 75.83


In [ ]:
# Calculating percentiles
percentile_age_25 = np.percentile(age, 25)  # 25th percentile
percentile_age_75 = np.percentile(age, 75)  # 75th percentile


In [ ]:
print(f"25th Percentile Age: {percentile_age_25:.2f}")
print(f"75th Percentile Age: {percentile_age_75:.2f}")


25th Percentile Age: 21.00
75th Percentile Age: 39.00


In [ ]:
# Calculating IQR
iqr_age = percentile_age_75 - percentile_age_25



In [ ]:
print(f"IQR Age: {iqr_age:.2f}")


IQR Age: 18.00


In [ ]:
# Calculating MAD (Median Absolute Deviation)
mad_age = median_abs_deviation(age, scale='normal')


In [ ]:
print(f"MAD Age: {mad_age:.2f}")


MAD Age: 11.86


From the above data and insights, we can see that -

- **Variance Age (201.11):** The variance is relatively high, indicating a significant spread in age values. This suggests diversity in the ages of passengers.

- **Standard Deviation Age (14.18):** A high standard deviation shows that ages are widely dispersed around the mean, reflecting a varied age distribution.

- **Range Age (75.83):** The large range indicates a broad spectrum of ages among the passengers, from the youngest to the oldest.

- **25th Percentile Age (21.00) and 75th Percentile Age (39.00):** These percentiles show that half of the passengers are between these ages, indicating a fairly consistent central age group.

- **Interquartile Range (IQR) Age (18.00):** The IQR provides a robust measure of dispersion, indicating that the central 50% of ages vary within this range, which is useful for understanding the typical spread of ages.

- **Mean Absolute Deviation (MAD) Age (11.86):** The MAD is lower than the standard deviation, suggesting that the median age is a more stable measure of central tendency compared to the mean in this context.



Thank you for reading! I’d love to hear how you plan to apply these concepts. Feel free to share your thoughts and experiences. Your journey into data analysis is just beginning, and I’m excited to explore more with you in the upcoming sections.

About Me <br>
Hi, I’m Krupa Dharamshi, an AIML Engineer with a passion for data science and machine learning. With a background in Artificial Intelligence and Machine Learning, I specialize in building intelligent systems and analyzing data to uncover valuable insights. I enjoy exploring new technologies and continuously learning to stay ahead in this ever-evolving field. Feel free to connect with me on [LinkedIn](https://www.linkedin.com/in/krupa-d-89233a226/) or explore my projects on [GitHub](https://github.com/krupa2002).